components/layout_2020/display_list/clip_path.rs

```diff
@@ -0,0 +1,228 @@
+use webrender_api::ClipChainId;
+use webrender_traits::display_list::ScrollTreeNodeId;
+
+use super::{compute_marginbox_radius, normalize_radii};
+
+pub(super) fn build(
+    clip_path: ClipPath,
+    display_list: &mut super::DisplayList,
+    parent_scroll_node_id: ScrollTreeNodeId,
+    parent_clip_chain_id: ClipChainId,
+    fragment_builder: super::BuilderForBoxFragment,
+) -> Option<ClipChainId> {
+    let geometry_box = match clip_path {
+        ClipPath::Shape(_, ShapeGeometryBox::ShapeBox(shape_box)) => shape_box,
+        ClipPath::Shape(_, ShapeGeometryBox::ElementDependent) => ShapeBox::BorderBox,
+        ClipPath::Box(ShapeGeometryBox::ShapeBox(shape_box)) => shape_box,
+        ClipPath::Box(ShapeGeometryBox::ElementDependent) => ShapeBox::BorderBox,
+        _ => return None,
+    };
+    let layout_rect = match geometry_box {
+        ShapeBox::BorderBox => fragment_builder.border_rect,
+        ShapeBox::ContentBox => *fragment_builder.content_rect(),
+        ShapeBox::PaddingBox => *fragment_builder.padding_rect(),
+        ShapeBox::MarginBox => *fragment_builder.margin_rect(),
+    };
+    if let ClipPath::Shape(shape, _) = clip_path {
+        match *shape {
+            BasicShape::Circle(_) | BasicShape::Ellipse(_) | BasicShape::Rect(_) => {
+                build_simple_shape(
+                    *shape,
+                    layout_rect,
+                    parent_scroll_node_id,
+                    parent_clip_chain_id,
+                    display_list,
+                )
+            },
+            BasicShape::Polygon(_) | BasicShape::PathOrShape(_) => None,
+        }
+    } else {
+        create_rect_clip_chain(
+            match geometry_box {
+                ShapeBox::MarginBox => compute_marginbox_radius(
+                    fragment_builder.border_radius,
+                    layout_rect.size(),
+                    fragment_builder.fragment,
+                ),
+                _ => fragment_builder.border_radius,
+            },
+            layout_rect,
+            parent_scroll_node_id,
+            parent_clip_chain_id,
+            display_list,
+        )
+    }
+}
+
+fn build_simple_shape(
+    shape: BasicShape,
+    layout: LayoutRect,
+    parent_scroll_node_id: ScrollTreeNodeId,
+    parent_clip_chain_id: ClipChainId,
+    display_list: &mut super::DisplayList,
+) -> Option<ClipChainId> {
+    match shape {
+        BasicShape::Rect(rect) => {
+            let top = rect.rect.0.resolve(Length::new(layout.height()));
+            let right = rect.rect.1.resolve(Length::new(layout.width()));
+            let bottom = rect.rect.2.resolve(Length::new(layout.height()));
+            let left = rect.rect.3.resolve(Length::new(layout.width()));
+            let x = layout.min.x + left.px();
+            let y = layout.min.y + top.px();
+            let width = layout.width() - (left + right).px();
+            let height = layout.height() - (top + bottom).px();
```

Comment:
I think this could be simplified slightly:

```suggestion
            let insets = LayoutSideOffsets::new(
                rect.rect.0.resolve(Length::new(layout.height())).px(),
                rect.rect.1.resolve(Length::new(layout.width())).px(),
                rect.rect.2.resolve(Length::new(layout.height())).px(),
                rect.rect.3.resolve(Length::new(layout.width())).px()
            );
            let shape_rect = layout.to_rect().inner_rect(insets).to_box2d();
```